File: frontend/src/types/customer.ts
export interface VendorDetail {
  name: string;
  service_provided?: string | null;
  notes?: string | null;
}

export interface QualityScore {
  metric_name: string;
  score: number;
  year?: number | null;
  source?: string | null;
}

export interface CustomerBase {
  name: string;
  description?: string | null;
  business_model?: string | null; // e.g., HMO, PPO, Value-Based Care
  membership_count?: number | null;

  website_url?: string | null; // Input will be string, validated by Pydantic on backend
  primary_contact_name?: string | null;
  primary_contact_email?: string | null;
  primary_contact_phone?: string | null;

  notes?: string | null;
}

export interface Customer extends CustomerBase {
  id: number;
  created_at: string; // Typically string in ISO format from backend
  updated_at: string; // Typically string in ISO format from backend
  quality_scores: QualityScore[];
  known_vendors: VendorDetail[];
}

export interface CustomerCreate extends CustomerBase {
  quality_scores?: QualityScore[];
  known_vendors?: VendorDetail[];
}

// For CustomerUpdate, we can often reuse parts of CustomerCreate or define it more flexibly
// if only partial updates are allowed and all fields are optional.
// Pydantic's Update schema makes all fields optional, so we'll mirror that.
export interface CustomerUpdate {
  name?: string | null;
  description?: string | null;
  business_model?: string | null;
  membership_count?: number | null;
  website_url?: string | null;
  primary_contact_name?: string | null;
  primary_contact_email?: string | null;
  primary_contact_phone?: string | null;
  notes?: string | null;
  quality_scores?: QualityScore[] | null;
  known_vendors?: VendorDetail[] | null;
}

// For forms, it's often useful to have a type that matches the form structure
export type CustomerFormData = Omit<CustomerCreate, 'quality_scores' | 'known_vendors'> & {
  // For complex nested forms, these might need their own form data types
  quality_scores: QualityScore[];
  known_vendors: VendorDetail[];
};

File: backend/app/schemas/customer.py
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, HttpUrl
import datetime

class VendorDetail(BaseModel):
    name: str
    service_provided: Optional[str] = None
    notes: Optional[str] = None

class QualityScore(BaseModel):
    metric_name: str
    score: float
    year: Optional[int] = None
    source: Optional[str] = None

class CustomerBase(BaseModel):
    name: str
    description: Optional[str] = None
    business_model: Optional[str] = None # e.g., HMO, PPO, Value-Based Care
    membership_count: Optional[int] = None

    website_url: Optional[HttpUrl] = None
    primary_contact_name: Optional[str] = None
    primary_contact_email: Optional[str] = None # Add email validation if needed
    primary_contact_phone: Optional[str] = None

    notes: Optional[str] = None

class CustomerCreate(CustomerBase):
    quality_scores: Optional[List[QualityScore]] = []
    known_vendors: Optional[List[VendorDetail]] = []

class CustomerUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    business_model: Optional[str] = None
    membership_count: Optional[int] = None
    website_url: Optional[HttpUrl] = None
    primary_contact_name: Optional[str] = None
    primary_contact_email: Optional[str] = None
    primary_contact_phone: Optional[str] = None
    notes: Optional[str] = None
    quality_scores: Optional[List[QualityScore]] = None
    known_vendors: Optional[List[VendorDetail]] = None

class Customer(CustomerBase):
    id: int
    created_at: datetime.datetime
    updated_at: datetime.datetime
    quality_scores: List[QualityScore] = []
    known_vendors: List[VendorDetail] = []

    class Config:
        orm_mode = True
